--- a/dolphin/src/settings/dolphin_generalsettings.kcfg
+++ b/dolphin/src/settings/dolphin_generalsettings.kcfg
@@ -56,6 +56,10 @@
             <label>Show selection toggle</label>
             <default>true</default>
         </entry>
+        <entry name="ShowKlookToggle" type="Bool">
+            <label>Show preview toggle</label>
+            <default>false</default>
+        </entry>
         <entry name="ShowToolTips" type="Bool">
             <label>Show tooltips</label>
             <default>false</default>
diff --git a/dolphin/src/settings/general/behaviorsettingspage.cpp b/dolphin/src/settings/general/behaviorsettingspage.cpp
index 1ade93c..0f3ed6f 100644
