--- a/dolphin/src/views/dolphinview.h
+++ b/dolphin/src/views/dolphinview.h
@@ -295,6 +295,8 @@ public:
      */
     void setViewPropertiesContext(const QString& context);
     QString viewPropertiesContext() const;
+    
+    KFileItemList sortedItems() const;
 
 public slots:
     /**
@@ -506,6 +508,8 @@ signals:
      */
     void writeStateChanged(bool isFolderWritable);
 
+    void klookTrigger();
+
     /**
      * Is emitted if the URL should be changed to the previous URL of the
      * history (e.g. because the "back"-mousebutton has been pressed).
