diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/CMakeLists.txt kde-baseapps-4.8.3/dolphin/src/CMakeLists.txt
--- kde-baseapps-4.8.3_old/dolphin/src/CMakeLists.txt	2012-06-01 14:42:38.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/CMakeLists.txt	2012-06-01 15:58:28.000000000 +0400
@@ -32,6 +32,7 @@
     kitemviews/kitemlistkeyboardsearchmanager.cpp
     kitemviews/kitemlistrubberband.cpp
     kitemviews/kitemlistselectionmanager.cpp
+    kitemviews/kitemlistklooktoggle.cpp
     kitemviews/kitemlistselectiontoggle.cpp
     kitemviews/kitemlistsizehintresolver.cpp
     kitemviews/kitemlistsmoothscroller.cpp
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/dolphinmainwindow.cpp kde-baseapps-4.8.3/dolphin/src/dolphinmainwindow.cpp
--- kde-baseapps-4.8.3_old/dolphin/src/dolphinmainwindow.cpp	2012-06-01 14:42:38.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/dolphinmainwindow.cpp	2012-06-01 15:58:28.000000000 +0400
@@ -2165,6 +2165,8 @@
             this, SLOT(enableStopAction()));
     connect(view, SIGNAL(finishedPathLoading(KUrl)),
             this, SLOT(disableStopAction()));
+    connect(view, SIGNAL(klookTrigger()),
+            this, SLOT(previewItem()));
 
     const KUrlNavigator* navigator = container->urlNavigator();
     connect(navigator, SIGNAL(urlChanged(KUrl)),
@@ -2354,12 +2356,20 @@
 
 void DolphinMainWindow::previewItem()
 {
-    const KFileItemList list = m_activeViewContainer->view()->selectedItems();
-    QStringList urlList;
+    KFileItemList list;
+    int index = 0;
+    if ( m_activeViewContainer->view()->selectedItemsCount() > 1) {
+        list = m_activeViewContainer->view()->selectedItems();
+    } else {
+        list = m_activeViewContainer->view()->sortedItems();
+        index = list.indexOf( m_activeViewContainer->view()->selectedItems().at(0));
+    }
+    QStringList urlList;    
     foreach(KFileItem item, list)
     {
         urlList.append(item.url().path());
-    }
+    }    
+    urlList.append("-i " + QString::number(index , 10));
     if (!list.isEmpty())
     {
         QProcess* process = new QProcess();
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kfileitemlistwidget.cpp kde-baseapps-4.8.3/dolphin/src/kitemviews/kfileitemlistwidget.cpp
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kfileitemlistwidget.cpp	2012-04-30 00:38:55.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kfileitemlistwidget.cpp	2012-06-01 15:58:28.000000000 +0400
@@ -164,7 +164,7 @@
     painter->setPen(Qt::red);
     painter->drawText(QPointF(0, itemListStyleOption.fontMetrics.height()), QString::number(index()));
     painter->drawRect(rect());
-#endif
+#endif    
 }
 
 QRectF KFileItemListWidget::iconRect() const
@@ -222,6 +222,48 @@
     return QRectF(pos, QSizeF(toggleSize, toggleSize));
 }
 
+QRectF KFileItemListWidget::klookToggleRect() const
+{
+    const_cast<KFileItemListWidget*>(this)->triggerCacheRefreshing();
+
+    const int iconHeight = styleOption().iconSize;
+
+    int toggleSize = KIconLoader::SizeSmall;
+    if (iconHeight >= KIconLoader::SizeEnormous) {
+        toggleSize = KIconLoader::SizeMedium;
+    } else if (iconHeight >= KIconLoader::SizeLarge) {
+        toggleSize = KIconLoader::SizeSmallMedium;
+    }
+
+    QPointF pos = QPointF(iconRect().topRight().rx() - toggleSize, iconRect().topRight().ry());
+
+    // If the selection toggle has a very small distance to the
+    // widget borders, the size of the selection toggle will get
+    // increased to prevent an accidental clicking of the item
+    // when trying to hit the toggle.
+
+    const int widgetHeight = size().height();
+    const int widgetWidth = size().width();
+    const int minMargin = 2;
+
+    if (toggleSize + minMargin * 2 >= widgetHeight) {
+        pos.rx() -= (widgetHeight - toggleSize) / 2;
+        toggleSize = widgetHeight;
+        pos.setY(0);
+    }
+    if (toggleSize + minMargin * 2 >= widgetWidth) {
+        pos.ry() -= (widgetWidth - toggleSize) / 2;
+        toggleSize = widgetWidth;
+        pos.setX(0);
+    }
+
+    return QRectF(pos, QSizeF(toggleSize, toggleSize));
+}
+
+
+
+
+
 QString KFileItemListWidget::roleText(const QByteArray& role, const QHash<QByteArray, QVariant>& values)
 {
     QString text;
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kfileitemlistwidget.h kde-baseapps-4.8.3/dolphin/src/kitemviews/kfileitemlistwidget.h
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kfileitemlistwidget.h	2012-03-30 00:25:30.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kfileitemlistwidget.h	2012-06-01 15:58:28.000000000 +0400
@@ -55,6 +55,7 @@
     virtual QRectF textRect() const;
     virtual QRectF expansionToggleRect() const;
     virtual QRectF selectionToggleRect() const;
+    virtual QRectF klookToggleRect() const;
 
     /**
      * @return Shown string for the role \p role of the item with the values \p values.
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistcontroller.cpp kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistcontroller.cpp
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistcontroller.cpp	2012-04-30 00:38:55.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistcontroller.cpp	2012-06-01 15:58:28.000000000 +0400
@@ -39,6 +39,7 @@
 
 #include <KGlobalSettings>
 #include <KDebug>
+#include <kapplication.h>
 
 KItemListController::KItemListController(QObject* parent) :
     QObject(parent),
@@ -484,6 +485,13 @@
         m_selectionManager->beginAnchoredSelection(m_pressedIndex);
         return true;
     }
+    if (m_view->isAboveKlookToggle(m_pressedIndex, m_pressedMousePos)){
+        m_selectionManager->setSelected(m_pressedIndex, 1, KItemListSelectionManager::Select);
+        m_selectionManager->setCurrentItem(m_pressedIndex);
+        m_selectionManager->beginAnchoredSelection(m_pressedIndex);
+        emit klookTrigger();
+        return true;
+    }
 
     const bool shiftPressed = event->modifiers() & Qt::ShiftModifier;
     const bool controlPressed = event->modifiers() & Qt::ControlModifier;
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistcontroller.h kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistcontroller.h
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistcontroller.h	2012-01-18 22:07:34.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistcontroller.h	2012-06-01 15:58:28.000000000 +0400
@@ -198,6 +198,8 @@
     void modelChanged(KItemModelBase* current, KItemModelBase* previous);
     void viewChanged(KItemListView* current, KItemListView* previous);
 
+    void klookTrigger();
+
 private slots:
     void slotViewScrollOffsetChanged(qreal current, qreal previous);
 
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistklooktoggle.cpp kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistklooktoggle.cpp
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistklooktoggle.cpp	1970-01-01 03:00:00.000000000 +0300
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistklooktoggle.cpp	2012-06-01 15:58:28.000000000 +0400
@@ -0,0 +1,106 @@
+/***************************************************************************
+ *   Copyright (C) 2011 by Peter Penz <peter.penz19@gmail.com>             *
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA            *
+ ***************************************************************************/
+
+#include "kitemlistklooktoggle_p.h"
+
+#include <KIconEffect>
+#include <KIconLoader>
+#include <QPainter>
+
+#include <KDebug>
+
+KItemListKlookToggle::KItemListKlookToggle(QGraphicsItem* parent) :
+    QGraphicsWidget(parent, 0),    
+    m_hovered(false)
+{
+    setAcceptHoverEvents(true);
+}
+
+KItemListKlookToggle::~KItemListKlookToggle()
+{
+}
+/*
+void KItemListKlookToggle::setChecked(bool checked)
+{
+    if (m_checked != checked) {
+        m_checked = checked;
+        m_pixmap = QPixmap();
+        update();
+    }
+}
+*/
+/*
+bool KItemListKlookToggle::isChecked() const
+{
+    return m_checked;
+}
+*/
+void KItemListKlookToggle::paint(QPainter* painter, const QStyleOptionGraphicsItem* option, QWidget* widget)
+{
+    Q_UNUSED(option);
+    Q_UNUSED(widget);
+
+    if (m_pixmap.isNull()) {
+        updatePixmap();
+    }
+
+    const qreal x = (size().width()  - qreal(m_pixmap.width()))  / 2;
+    const qreal y = (size().height() - qreal(m_pixmap.height())) / 2;
+    painter->drawPixmap(x, y, m_pixmap);
+}
+
+void KItemListKlookToggle::hoverEnterEvent(QGraphicsSceneHoverEvent* event)
+{
+    QGraphicsWidget::hoverEnterEvent(event);
+    m_hovered = true;
+    m_pixmap = QPixmap();
+}
+
+void KItemListKlookToggle::hoverLeaveEvent(QGraphicsSceneHoverEvent* event)
+{
+    QGraphicsWidget::hoverLeaveEvent(event);
+    m_hovered = false;
+    m_pixmap = QPixmap();
+}
+
+void KItemListKlookToggle::updatePixmap()
+{
+    const char* icon = "klook";
+
+    int iconSize = qMin(size().width(), size().height());
+    if (iconSize < KIconLoader::SizeSmallMedium) {
+        iconSize = KIconLoader::SizeSmall;
+    } else if (iconSize < KIconLoader::SizeMedium) {
+        iconSize = KIconLoader::SizeSmallMedium;
+    } else if (iconSize < KIconLoader::SizeLarge) {
+        iconSize = KIconLoader::SizeMedium;
+    } else if (iconSize < KIconLoader::SizeHuge) {
+        iconSize = KIconLoader::SizeLarge;
+    } else if (iconSize < KIconLoader::SizeEnormous) {
+        iconSize = KIconLoader::SizeHuge;
+    }
+
+    m_pixmap = KIconLoader::global()->loadIcon(QLatin1String(icon), KIconLoader::NoGroup, iconSize);
+
+    if (m_hovered) {
+        KIconLoader::global()->iconEffect()->apply(m_pixmap, KIconLoader::Desktop, KIconLoader::ActiveState);
+    }
+}
+
+#include "kitemlistklooktoggle_p.moc"
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistklooktoggle_p.h kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistklooktoggle_p.h
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistklooktoggle_p.h	1970-01-01 03:00:00.000000000 +0300
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistklooktoggle_p.h	2012-06-01 15:58:28.000000000 +0400
@@ -0,0 +1,61 @@
+/***************************************************************************
+ *   Copyright (C) 2011 by Peter Penz <peter.penz19@gmail.com>             *
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program; if not, write to the                         *
+ *   Free Software Foundation, Inc.,                                       *
+ *   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA            *
+ ***************************************************************************/
+
+#ifndef KITEMLISTKLOOKTOGGLE_H
+#define KITEMLISTKLOOKTOGGLE_H
+
+#include <libdolphin_export.h>
+
+#include <QGraphicsWidget>
+#include <QPixmap>
+
+class QPropertyAnimation;
+
+/**
+ * @brief Allows to launch klook by clicking on icon overlay
+ */
+class LIBDOLPHINPRIVATE_EXPORT KItemListKlookToggle : public QGraphicsWidget
+{
+    Q_OBJECT
+
+public:
+    KItemListKlookToggle(QGraphicsItem* parent);
+    virtual ~KItemListKlookToggle();
+
+    //void setChecked(bool checked);
+    //bool isChecked() const;
+
+    virtual void paint(QPainter* painter, const QStyleOptionGraphicsItem* option, QWidget* widget = 0);
+
+protected:
+    virtual void hoverEnterEvent(QGraphicsSceneHoverEvent* event);
+    virtual void hoverLeaveEvent(QGraphicsSceneHoverEvent* event);
+
+private:
+    void updatePixmap();
+
+private:
+    //bool m_checked;
+    bool m_hovered;
+    QPixmap m_pixmap;
+};
+
+#endif
+
+
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistview.cpp kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistview.cpp
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistview.cpp	2012-04-30 00:38:55.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistview.cpp	2012-06-01 15:58:28.000000000 +0400
@@ -53,6 +53,7 @@
 KItemListView::KItemListView(QGraphicsWidget* parent) :
     QGraphicsWidget(parent),
     m_enabledSelectionToggles(false),
+    m_enabledKlookToggles(false),
     m_grouped(false),
     m_supportsItemExpanding(false),
     m_activeTransactions(0),
@@ -320,6 +321,25 @@
     return m_enabledSelectionToggles;
 }
 
+void KItemListView::setEnabledKlookToggles(bool enabled)
+{
+    if (m_enabledKlookToggles != enabled) {
+        m_enabledKlookToggles = enabled;
+
+        QHashIterator<int, KItemListWidget*> it(m_visibleItems);
+        while (it.hasNext()) {
+            it.next();
+            it.value()->setEnabledKlookToggle(enabled);
+        }
+    }
+}
+
+bool KItemListView::enabledKlookToggles() const
+{
+    return m_enabledKlookToggles;
+}
+
+
 KItemListController* KItemListView::controller() const
 {
     return m_controller;
@@ -497,6 +517,25 @@
     return false;
 }
 
+bool KItemListView::isAboveKlookToggle(int index, const QPointF& pos) const
+{
+
+    if (!m_enabledKlookToggles) {
+        return false;
+    }
+
+    const KItemListWidget* widget = m_visibleItems.value(index);
+    if (widget) {
+        const QRectF selectionToggleRect = widget->klookToggleRect();
+        if (!selectionToggleRect.isEmpty()) {
+            const QPointF mappedPos = widget->mapFromItem(this, pos);
+            return selectionToggleRect.contains(mappedPos);
+        }
+    }
+    return false;
+}
+
+
 int KItemListView::firstVisibleIndex() const
 {
     return m_layouter->firstVisibleIndex();
@@ -1757,6 +1796,7 @@
     widget->setSelected(selectionManager->isSelected(index));
     widget->setHovered(false);
     widget->setEnabledSelectionToggle(enabledSelectionToggles());
+    widget->setEnabledKlookToggle(enabledKlookToggles());
     widget->setIndex(index);
     widget->setData(m_model->data(index));
     widget->setSiblingsInformation(QBitArray());
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistview.h kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistview.h
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistview.h	2012-03-30 00:25:30.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistview.h	2012-06-01 15:58:28.000000000 +0400
@@ -122,6 +122,9 @@
     void setEnabledSelectionToggles(bool enabled);
     bool enabledSelectionToggles() const;
 
+    void setEnabledKlookToggles(bool enabled);
+    bool enabledKlookToggles() const;
+
     /**
      * @return Controller of the item-list. The controller gets
      *         initialized by KItemListController::setView() and will
@@ -171,6 +174,7 @@
     bool isAboveSelectionToggle(int index, const QPointF& pos) const;
     bool isAboveExpansionToggle(int index, const QPointF& pos) const;
 
+    bool isAboveKlookToggle(int index, const QPointF& pos) const;
     /**
      * @return Index of the first item that is at least partly visible.
      *         -1 is returned if the model contains no items.
@@ -625,6 +629,7 @@
 
 private:
     bool m_enabledSelectionToggles;
+    bool m_enabledKlookToggles;
     bool m_grouped;
     bool m_supportsItemExpanding;
     int m_activeTransactions; // Counter for beginTransaction()/endTransaction()
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistwidget.cpp kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistwidget.cpp
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistwidget.cpp	2012-03-01 02:56:09.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistwidget.cpp	2012-06-01 15:58:28.000000000 +0400
@@ -23,6 +23,7 @@
 #include "kitemlistwidget.h"
 
 #include "kitemlistselectiontoggle_p.h"
+#include "kitemlistklooktoggle_p.h"
 #include "kitemlistview.h"
 #include "kitemmodelbase.h"
 
@@ -42,6 +43,7 @@
     m_hovered(false),
     m_alternateBackground(false),
     m_enabledSelectionToggle(false),
+    m_enabledKlookToggle(false),
     m_data(),
     m_visibleRoles(),
     m_visibleRolesSizes(),
@@ -50,7 +52,8 @@
     m_hoverOpacity(0),
     m_hoverCache(0),
     m_hoverAnimation(0),
-    m_selectionToggle(0)
+    m_selectionToggle(0),
+    m_klookToggle(0)
 {
 }
 
@@ -65,6 +68,9 @@
         delete m_selectionToggle;
         m_selectionToggle = 0;
 
+        delete m_klookToggle;
+        m_klookToggle = 0;
+
         if (m_hoverAnimation) {
             m_hoverAnimation->stop();
             m_hoverOpacity = 0;
@@ -260,6 +266,9 @@
         if (m_enabledSelectionToggle && !(QApplication::mouseButtons() & Qt::LeftButton)) {
             initializeSelectionToggle();
         }
+        if (m_enabledKlookToggle && !(QApplication::mouseButtons() & Qt::LeftButton)) {
+            initializeKlookToggle();
+        }
     } else {
         m_hoverAnimation->setStartValue(hoverOpacity());
         m_hoverAnimation->setEndValue(0.0);
@@ -291,7 +300,7 @@
 }
 
 void KItemListWidget::setEnabledSelectionToggle(bool enable)
-{
+{    
     if (m_enabledSelectionToggle != enable) {
         m_enabledSelectionToggle = enable;
         update();
@@ -303,6 +312,19 @@
     return m_enabledSelectionToggle;
 }
 
+void KItemListWidget::setEnabledKlookToggle(bool enable)
+{ 
+    if (m_enabledKlookToggle != enable) {
+        m_enabledKlookToggle = enable;
+        update();
+    }
+}
+
+bool KItemListWidget::enabledKlookToggle() const
+{
+    return m_enabledKlookToggle;
+}
+
 void KItemListWidget::setSiblingsInformation(const QBitArray& siblings)
 {
     const QBitArray previous = m_siblingsInfo;
@@ -333,6 +355,12 @@
     return QRectF();
 }
 
+QRectF KItemListWidget::klookToggleRect() const
+{
+    return QRectF();
+}
+
+
 QRectF KItemListWidget::expansionToggleRect() const
 {
     return QRectF();
@@ -408,6 +436,8 @@
     if (!m_hovered) {
         delete m_selectionToggle;
         m_selectionToggle = 0;
+        delete m_klookToggle;
+        m_klookToggle = 0;
     }
 }
 
@@ -426,12 +456,29 @@
     m_selectionToggle->setChecked(isSelected());
 }
 
+void KItemListWidget::initializeKlookToggle()
+{
+    Q_ASSERT(m_enabledKlookToggle);
+
+    if (!m_klookToggle) {
+        m_klookToggle = new KItemListKlookToggle(this);
+    }
+
+    const QRectF toggleRect = klookToggleRect();;
+    m_klookToggle->setPos(toggleRect.topLeft());
+    m_klookToggle->resize(toggleRect.size());
+
+}
+
 void KItemListWidget::setHoverOpacity(qreal opacity)
 {
     m_hoverOpacity = opacity;
     if (m_selectionToggle) {
         m_selectionToggle->setOpacity(opacity);
     }
+    if (m_klookToggle) {
+        m_klookToggle->setOpacity(opacity);
+    }
 
     if (m_hoverOpacity <= 0.0) {
         delete m_hoverCache;
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistwidget.h kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistwidget.h
--- kde-baseapps-4.8.3_old/dolphin/src/kitemviews/kitemlistwidget.h	2012-03-01 02:56:09.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/kitemviews/kitemlistwidget.h	2012-06-01 15:58:28.000000000 +0400
@@ -32,6 +32,7 @@
 #include <QStyle>
 
 class KItemListSelectionToggle;
+class KItemListKlookToggle;
 class QPropertyAnimation;
 
 /**
@@ -88,6 +89,9 @@
     void setEnabledSelectionToggle(bool enabled);
     bool enabledSelectionToggle() const;
 
+    void setEnabledKlookToggle(bool enable);
+    bool enabledKlookToggle() const;
+
     /**
      * Sets the sibling information for the item and all of its parents.
      * The sibling information of the upper most parent is represented by
@@ -130,6 +134,11 @@
      */
     virtual QRectF expansionToggleRect() const;
 
+    /**
+     * @return Rectangle for the klook-toggle that is used for item preview.
+     */
+    virtual QRectF klookToggleRect() const;
+
 protected:
     virtual void dataChanged(const QHash<QByteArray, QVariant>& current, const QSet<QByteArray>& roles = QSet<QByteArray>());
     virtual void visibleRolesChanged(const QList<QByteArray>& current, const QList<QByteArray>& previous);
@@ -153,6 +162,7 @@
 
 private:
     void initializeSelectionToggle();
+    void initializeKlookToggle();
     void setHoverOpacity(qreal opacity);
     void clearHoverCache();
     void drawItemStyleOption(QPainter* painter, QWidget* widget, QStyle::State styleState);
@@ -166,6 +176,7 @@
     bool m_hovered;
     bool m_alternateBackground;
     bool m_enabledSelectionToggle;
+    bool m_enabledKlookToggle;
     QHash<QByteArray, QVariant> m_data;
     QList<QByteArray> m_visibleRoles;
     QHash<QByteArray, QSizeF> m_visibleRolesSizes;
@@ -177,6 +188,7 @@
     QPropertyAnimation* m_hoverAnimation;
 
     KItemListSelectionToggle* m_selectionToggle;
+    KItemListKlookToggle* m_klookToggle;
 };
 #endif
 
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/settings/dolphin_generalsettings.kcfg kde-baseapps-4.8.3/dolphin/src/settings/dolphin_generalsettings.kcfg
--- kde-baseapps-4.8.3_old/dolphin/src/settings/dolphin_generalsettings.kcfg	2012-04-30 00:38:55.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/settings/dolphin_generalsettings.kcfg	2012-06-01 15:58:28.000000000 +0400
@@ -56,6 +56,10 @@
             <label>Show selection toggle</label>
             <default>true</default>
         </entry>
+        <entry name="ShowKlookToggle" type="Bool">
+            <label>Show preview toggle</label>
+            <default>false</default>
+        </entry>
         <entry name="ShowToolTips" type="Bool">
             <label>Show tooltips</label>
             <default>false</default>
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/settings/general/behaviorsettingspage.cpp kde-baseapps-4.8.3/dolphin/src/settings/general/behaviorsettingspage.cpp
--- kde-baseapps-4.8.3_old/dolphin/src/settings/general/behaviorsettingspage.cpp	2011-12-03 01:17:41.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/settings/general/behaviorsettingspage.cpp	2012-06-01 15:58:28.000000000 +0400
@@ -46,7 +46,8 @@
     m_confirmDelete(0),
     m_showToolTips(0),
     m_showSelectionToggle(0),
-    m_naturalSorting(0)
+    m_naturalSorting(0),
+    m_showKlookToggle(0)
 {
     QVBoxLayout* topLayout = new QVBoxLayout(this);
 
@@ -86,10 +87,14 @@
     // 'Natural sorting of items'
     m_naturalSorting = new QCheckBox(i18nc("option:check", "Natural sorting of items"), this);
 
+    // 'Show klook toggle'
+    m_showKlookToggle = new QCheckBox(i18nc("@option:check", "Show preview marker"), this);
+
     topLayout->addWidget(propsBox);
     topLayout->addWidget(confirmBox);
     topLayout->addWidget(m_showToolTips);
     topLayout->addWidget(m_showSelectionToggle);
+    topLayout->addWidget(m_showKlookToggle);
     topLayout->addWidget(m_naturalSorting);
     topLayout->addStretch();
 
@@ -102,6 +107,7 @@
     connect(m_confirmClosingMultipleTabs, SIGNAL(toggled(bool)), this, SIGNAL(changed()));
     connect(m_showToolTips, SIGNAL(toggled(bool)), this, SIGNAL(changed()));
     connect(m_showSelectionToggle, SIGNAL(toggled(bool)), this, SIGNAL(changed()));
+    connect(m_showKlookToggle, SIGNAL(toggled(bool)), this, SIGNAL(changed()));
     connect(m_naturalSorting, SIGNAL(toggled(bool)), this, SIGNAL(changed()));
 }
 
@@ -136,6 +142,7 @@
     settings->setConfirmClosingMultipleTabs(m_confirmClosingMultipleTabs->isChecked());
     settings->setShowToolTips(m_showToolTips->isChecked());
     settings->setShowSelectionToggle(m_showSelectionToggle->isChecked());
+    settings->setShowKlookToggle(m_showKlookToggle->isChecked());
     settings->writeConfig();
 
     const bool naturalSorting = m_naturalSorting->isChecked();
@@ -172,6 +179,7 @@
     m_confirmClosingMultipleTabs->setChecked(GeneralSettings::confirmClosingMultipleTabs());
     m_showToolTips->setChecked(GeneralSettings::showToolTips());
     m_showSelectionToggle->setChecked(GeneralSettings::showSelectionToggle());
+    m_showKlookToggle->setChecked(GeneralSettings::showKlookToggle());
     m_naturalSorting->setChecked(KGlobalSettings::naturalSorting());
 }
 
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/settings/general/behaviorsettingspage.h kde-baseapps-4.8.3/dolphin/src/settings/general/behaviorsettingspage.h
--- kde-baseapps-4.8.3_old/dolphin/src/settings/general/behaviorsettingspage.h	2011-12-03 01:17:41.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/settings/general/behaviorsettingspage.h	2012-06-01 15:58:28.000000000 +0400
@@ -61,6 +61,7 @@
     QLabel* m_configureToolTips;
     QCheckBox* m_showSelectionToggle;
     QCheckBox* m_naturalSorting;
+    QCheckBox* m_showKlookToggle;
 };
 
 #endif
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/views/dolphinitemlistcontainer.cpp kde-baseapps-4.8.3/dolphin/src/views/dolphinitemlistcontainer.cpp
--- kde-baseapps-4.8.3_old/dolphin/src/views/dolphinitemlistcontainer.cpp	2012-04-30 00:38:55.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/views/dolphinitemlistcontainer.cpp	2012-06-01 15:58:28.000000000 +0400
@@ -52,6 +52,7 @@
 
     m_fileItemListView->setWidgetCreator(new KItemListWidgetCreator<DolphinFileItemListWidget>());
     m_fileItemListView->setEnabledSelectionToggles(GeneralSettings::showSelectionToggle());
+    m_fileItemListView->setEnabledKlookToggles(GeneralSettings::showKlookToggle());
     m_fileItemListView->setEnlargeSmallPreviews(GeneralSettings::enlargeSmallPreviews());
 
     updateAutoActivationDelay();
@@ -181,6 +182,7 @@
     beginTransaction();
 
     m_fileItemListView->setEnabledSelectionToggles(GeneralSettings::showSelectionToggle());
+    m_fileItemListView->setEnabledKlookToggles(GeneralSettings::showKlookToggle());
 
     const bool expandableFolders = (itemLayout() && KFileItemListView::DetailsLayout) &&
                                    DetailsModeSettings::expandableFolders();
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/views/dolphinview.cpp kde-baseapps-4.8.3/dolphin/src/views/dolphinview.cpp
--- kde-baseapps-4.8.3_old/dolphin/src/views/dolphinview.cpp	2012-06-01 14:42:38.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/views/dolphinview.cpp	2012-06-01 15:58:28.000000000 +0400
@@ -157,6 +157,7 @@
     connect(controller, SIGNAL(itemUnhovered(int)), this, SLOT(slotItemUnhovered(int)));
     connect(controller, SIGNAL(itemDropEvent(int,QGraphicsSceneDragDropEvent*)), this, SLOT(slotItemDropEvent(int,QGraphicsSceneDragDropEvent*)));
     connect(controller, SIGNAL(modelChanged(KItemModelBase*,KItemModelBase*)), this, SLOT(slotModelChanged(KItemModelBase*,KItemModelBase*)));
+    connect(controller, SIGNAL(klookTrigger()), this, SIGNAL(klookTrigger()));
 
     KFileItemModel* model = fileItemModel();
     if (model) {
@@ -1482,4 +1483,15 @@
     return sortHash.value(sortRole);
 }
 
+KFileItemList DolphinView::sortedItems() const
+{
+    KFileItemModel*  model = fileItemModel();
+    KFileItemList  list;
+    for (int i = 0; i < model->count(); i++) {
+        list.append(model->fileItem(i));
+    }
+    return list;
+}
+
+
 #include "dolphinview.moc"
diff -Nuar kde-baseapps-4.8.3_old/dolphin/src/views/dolphinview.h kde-baseapps-4.8.3/dolphin/src/views/dolphinview.h
--- kde-baseapps-4.8.3_old/dolphin/src/views/dolphinview.h	2012-06-01 14:42:38.000000000 +0400
+++ kde-baseapps-4.8.3/dolphin/src/views/dolphinview.h	2012-06-01 16:02:39.000000000 +0400
@@ -350,7 +350,8 @@
      */
     static bool isFileOnRemovableMedia(const KUrl::List& urlList);
 
-
+    KFileItemList sortedItems() const;
+    
 public slots:
     /**
      * Changes the directory to \a url. If the current directory is equal to
@@ -557,6 +558,8 @@
      */
     void writeStateChanged(bool isFolderWritable);
 
+    void klookTrigger();
+
 protected:
     /** Changes the zoom level if Control is pressed during a wheel event. */
     virtual void wheelEvent(QWheelEvent* event);
